File: src/mcp-server/storage/rule_content.py
"""
Rule content manager for AutoImprove.

Manages individual rule content files (rules/content/rule-{id}.md).
"""

from pathlib import Path
from typing import Optional, Dict, Any
import frontmatter


class RuleContent:
    """Represents a rule's full content."""

    def __init__(
        self,
        id: str,
        content: str,
        reason: str,
        metadata: Dict[str, Any]
    ):
        """
        Initialize rule content.

        Args:
            id: Rule ID
            content: Rule content (what to do)
            reason: Why this rule exists
            metadata: Additional metadata (type, priority, scenes, etc.)
        """
        self.id = id
        self.content = content
        self.reason = reason
        self.metadata = metadata

    def to_markdown(self) -> str:
        """
        Convert to markdown format with frontmatter.

        Returns:
            Markdown string with YAML frontmatter
        """
        post = frontmatter.Post(
            f"## Content\n\n{self.content}\n\n## Reason\n\n{self.reason}",
            **self.metadata
        )
        return frontmatter.dumps(post)

    @classmethod
    def from_markdown(cls, rule_id: str, markdown: str) -> "RuleContent":
        """
        Parse from markdown format.

        Args:
            rule_id: Rule ID
            markdown: Markdown string with frontmatter

        Returns:
            RuleContent instance
        """
        post = frontmatter.loads(markdown)

        # Extract content and reason from body
        body = post.content
        content = ""
        reason = ""

        if "## Content" in body and "## Reason" in body:
            parts = body.split("## Reason")
            content = parts[0].replace("## Content", "").strip()
            reason = parts[1].strip()
        else:
            content = body

        return cls(
            id=rule_id,
            content=content,
            reason=reason,
            metadata=post.metadata
        )


class RuleContentManager:
    """Manages rule content file operations."""

    def __init__(self, storage_root: Path):
        """
        Initialize rule content manager.

        Args:
            storage_root: Path to ~/.autoimprove/
        """
        self.storage_root = storage_root
        self.content_dir = storage_root / "rules" / "content"

    def _get_content_path(self, rule_id: str) -> Path:
        """Get path to rule content file."""
        return self.content_dir / f"{rule_id}.md"

    def load_content(self, rule_id: str) -> Optional[RuleContent]:
        """
        Load rule content from disk.

        Args:
            rule_id: ID of rule to load

        Returns:
            RuleContent if found, None otherwise
        """
        path = self._get_content_path(rule_id)

        if not path.exists():
            return None

        with open(path) as f:
            markdown = f.read()

        return RuleContent.from_markdown(rule_id, markdown)

    def save_content(self, rule: RuleContent) -> None:
        """
        Save rule content to disk atomically.

        Args:
            rule: RuleContent to save
        """
        # Ensure directory exists
        self.content_dir.mkdir(parents=True, exist_ok=True)

        path = self._get_content_path(rule.id)

        # Write to temp file first
        temp_path = path.with_suffix('.md.tmp')
        with open(temp_path, 'w') as f:
            f.write(rule.to_markdown())

        # Atomic rename
        temp_path.replace(path)

    def delete_content(self, rule_id: str) -> bool:
        """
        Delete rule content file.

        Args:
            rule_id: ID of rule to delete

        Returns:
            True if deleted, False if not found
        """
        path = self._get_content_path(rule_id)

        if not path.exists():
            return False

        path.unlink()
        return True

    def exists(self, rule_id: str) -> bool:
        """
        Check if rule content file exists.

        Args:
            rule_id: ID of rule to check

        Returns:
            True if exists, False otherwise
        """
        return self._get_content_path(rule_id).exists()

File: src/mcp-server-ts/src/storage/rule-content.ts
/**
 * Rule content manager for AutoImprove.
 *
 * Manages individual rule content files (rules/content/rule-{id}.md).
 */

import { readFileSync, writeFileSync, unlinkSync, existsSync, mkdirSync, renameSync } from "fs";
import { join } from "path";
import { homedir } from "os";
import { RuleContent } from "../core/models.js";

// Simple frontmatter parser/serializer
function parseFrontmatter(markdown: string): { metadata: Record<string, any>; content: string } {
  const lines = markdown.split("\n");

  if (lines[0] !== "---") {
    return { metadata: {}, content: markdown };
  }

  let endIndex = -1;
  for (let i = 1; i < lines.length; i++) {
    if (lines[i] === "---") {
      endIndex = i;
      break;
    }
  }

  if (endIndex === -1) {
    return { metadata: {}, content: markdown };
  }

  // Parse YAML-like frontmatter
  const metadataLines = lines.slice(1, endIndex);
  const metadata: Record<string, any> = {};

  for (const line of metadataLines) {
    const match = line.match(/^(\w+):\s*(.+)$/);
    if (match) {
      const [, key, value] = match;
      // Try to parse as JSON, fallback to string
      try {
        metadata[key] = JSON.parse(value);
      } catch {
        metadata[key] = value;
      }
    }
  }

  const content = lines.slice(endIndex + 1).join("\n").trim();

  return { metadata, content };
}

function serializeFrontmatter(metadata: Record<string, any>, content: string): string {
  const lines = ["---"];

  for (const [key, value] of Object.entries(metadata)) {
    const serialized = typeof value === "string" ? value : JSON.stringify(value);
    lines.push(`${key}: ${serialized}`);
  }

  lines.push("---");
  lines.push("");
  lines.push(content);

  return lines.join("\n");
}

export class RuleContentManager {
  private getStorageRoot(): string {
    return process.env.AUTOIMPROVE_STORAGE_ROOT || join(homedir(), ".autoimprove");
  }

  private getContentDir(): string {
    return join(this.getStorageRoot(), "rules", "content");
  }

  constructor() {
    this.ensureDirectory();
  }

  private ensureDirectory(): void {
    const contentDir = this.getContentDir();
    if (!existsSync(contentDir)) {
      mkdirSync(contentDir, { recursive: true });
    }
  }

  private getContentPath(ruleId: string): string {
    return join(this.getContentDir(), `${ruleId}.md`);
  }

  loadContent(ruleId: string): RuleContent | null {
    const path = this.getContentPath(ruleId);

    if (!existsSync(path)) {
      return null;
    }

    const markdown = readFileSync(path, "utf-8");
    const { metadata, content: body } = parseFrontmatter(markdown);

    // Extract content and reason from body
    let content = "";
    let reason = "";

    if (body.includes("## Content") && body.includes("## Reason")) {
      const parts = body.split("## Reason");
      content = parts[0].replace("## Content", "").trim();
      reason = parts[1].trim();
    } else {
      content = body;
    }

    return {
      id: ruleId,
      content,
      reason,
      metadata
    };
  }

  saveContent(rule: RuleContent): void {
    this.ensureDirectory();

    const path = this.getContentPath(rule.id);
    const body = `## Content\n\n${rule.content}\n\n## Reason\n\n${rule.reason}`;
    const markdown = serializeFrontmatter(rule.metadata, body);

    // Write to temp file first for atomic operation
    const tempPath = path + ".tmp";
    writeFileSync(tempPath, markdown);

    // Atomic rename
    renameSync(tempPath, path);
  }

  deleteContent(ruleId: string): boolean {
    const path = this.getContentPath(ruleId);

    if (!existsSync(path)) {
      return false;
    }

    unlinkSync(path);
    return true;
  }

  exists(ruleId: string): boolean {
    return existsSync(this.getContentPath(ruleId));
  }

  toMarkdown(rule: RuleContent): string {
    const body = `## Content\n\n${rule.content}\n\n## Reason\n\n${rule.reason}`;
    return serializeFrontmatter(rule.metadata, body);
  }
}
